Add recommendations to allow for users to recommend others on their profile

Pritz395 opened the pull request:
## Summary
Implements a comprehensive user recommendations feature allowing users to write, request, manage, and display recommendations on profiles. All changes are **additive only** and do not modify any pre-existing BLT/main codebase functionality.

## Features Implemented

### Core Recommendation System
- Write recommendations with relationship type, skills endorsement (up to 5), and text (200-1000 chars)
- Approval workflow: Recommendations require recipient approval before being visible
- One recommendation per user pair (enforced via unique constraint)
- Self-recommendation prevention
- Rate limiting (5 recommendations per day per user)

### LinkedIn-Style Enhancements
- **Request Recommendation**: Users can request recommendations from others with optional messages
- **Edit Pending Recommendations**: Recommenders can edit pending recommendations before approval
- **Recommendation Blurb**: Profile owners can add a short "About" section (max 500 chars) for recommendations
- **Highlight Recommendations**: Recipients can pin up to 3 recommendations to show first with visual distinction

### Request Management
- Track sent requests (pending, accepted, declined, completed, cancelled)
- Track received requests with accept/decline options
- Cancel pending requests (sender) with 2FA confirmation
- Request status updates automatically when recommendation is written

### UI/UX Improvements
- Modern form design with gradient headers and improved styling
- Skills selection via checkboxes (grouped by category) instead of dropdown
- Real-time character counters and skill selection limits
- Three-tab interface: "Received" (green), "Given" (green), "Pending" (blue)
- "See more/less" functionality for long recommendation text
- Double confirmation (2FA-like) for critical actions (approve, delete, cancel)
- Recommendation blurb displayed prominently in profile sidebar

### Database & Migrations
- Consolidated from 9 migration files to 3 essential migrations
- Removed all testing/temporary migrations
- Clean migration history for easier maintainer review

### Testing
- Comprehensive Django test suite (15 test cases)
- Tests cover: creation, approval, requests, editing, highlighting, security validations
- All tests passing

## Technical Details

### Models Added
- `Recommendation` - Core recommendation model with approval workflow
- `RecommendationRequest` - Tracks recommendation requests between users
- `RecommendationSkill` - Skills that can be endorsed
- `UserProfile.recommendation_blurb` - Short profile summary field

### Views Added
- `add_recommendation` - Create new recommendations
- `approve_recommendation` - Approve/reject pending recommendations
- `delete_recommendation` - Delete pending recommendations
- `request_recommendation` - Send recommendation requests
- `respond_to_request` - Accept/decline/cancel requests
- `edit_recommendation` - Edit pending recommendations
- `highlight_recommendation` - Pin/unpin recommendations
- `edit_recommendation_blurb` - Manage profile blurb

### Security & Validation
- Authentication required for all actions
- Authorization checks (only recipient can approve, only recommender can edit pending)
- Self-recommendation prevention
- Duplicate recommendation prevention (unique constraint)
- Rate limiting (5 recommendations/day)
- Character limits enforced (200-1000 for text, 500 for blurb/messages)
- Maximum 5 skills per recommendation
- 2FA-style double confirmation for irreversible actions
- Approved recommendations cannot be deleted (solidified)

## Files Changed
- **Models**: `website/models.py` (additive only)
- **Forms**: `website/forms.py` (new forms added)
- **Views**: `website/views/user.py` (new views added, existing views extended)
- **Templates**: New templates + `profile.html` updates (additive)
- **URLs**: `blt/urls.py` (new routes added)
- **Admin**: `website/admin.py` (new admin registrations)
- **Migrations**: 3 consolidated migration files
- **Tests**: `website/tests/test_recommendations.py` (15 test cases)

## Migration Notes
- Migrations consolidated from 9 files to 3 for cleaner history
- All testing/temporary migrations removed
- Ready for production deployment

## Related Issue
Closes #2632

## Checklist
- [x] All changes are additive (no modifications to pre-existing features)
- [x] Pre-commit hooks passing (isort, ruff, ruff-format, djlint)
- [x] All tests passing
- [x] Migrations consolidated and clean
- [x] Error handling and validation in place
- [x] Security checks implemented
- [x] UI/UX polished and responsive
- [x] Documentation removed (internal .md files cleaned up)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Full recommendations workflow on profiles: write, request, respond, approve, edit, delete, highlight, and list recommendations; profile “About” (recommendation blurb).

* **UX**
  * New profile UI: Received/Given/Pending tabs, modal views, counts, CTAs; live character counters (200–1000 for recommendations, 0–500 for requests), 5-skill selection limit, client-side validation, and confirmation prompts.

* **Admin**
  * Admin interfaces to manage recommendations, requests, and skills.

* **Tests**
  * Comprehensive test suite covering recommendation flows, validations, and UI behaviors.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Pritz395 committed:
feat: Add user recommendations feature with LinkedIn-style enhancements

Pritz395 committed:
fix: Apply pre-commit formatting and linting fixes

Pritz395 committed:
fix: Improve recommendation blurb workflow and positioning

Pritz395 committed:
feat: Improve recommendation blurb UI/UX and move to main content area

Pritz395 committed:
feat: Improve recommendation forms UI/UX with modern design

Pritz395 committed:
feat: Replace skills dropdown with checkboxes for better UX

Pritz395 committed:
fix: Update save method to handle checkbox values correctly

Pritz395 committed:
refactor: Consolidate migrations - reduce from 9 to 3 files

- Removed testing migrations (0260-0263, 0266)
- Merged 0265 into 0264 (cancelled status included from start)
- Final migrations: 0258, 0259, 0264

Pritz395 committed:
test: Add comprehensive Django tests for recommendation feature

Pritz395 committed:
fix: Remove exception swallowing and invalid skip_validation kwarg from Recommendation.save

Pritz395 committed:
fix: Change const to let for count variable in recommendation form to allow reassignment

Pritz395 committed:
fix: Add transaction.atomic() and select_for_update() to prevent race condition in highlight_recommendation

Pritz395 committed:
fix: Add height/width attributes to img tags and fix djlint orphan tag warning

Pritz395 committed:
fix: Address CodeRabbit review comments - use tuples for choices/Meta, fix help_text, preserve has_recommended, safe avatar lookup

Pritz395 commented:
@DonnieBLT There's a pre-existing issue, I need help with the tests. I think the run-test is hanging?

Pritz395 commented:
@coderabbitai what's causing the run and pre-commit test(s) from failing?

DonnieBLT commented on blt/urls.py:
```diff
@@ -573,6 +582,18 @@
+    path("recommendations/add/<str:username>/", add_recommendation, name="add_recommendation"),
+    path("recommendations/<int:recommendation_id>/approve/", approve_recommendation, name="approve_recommendation"),
+    path("recommendations/<int:recommendation_id>/delete/", delete_recommendation, name="delete_recommendation"),
+    path("recommendations/<int:recommendation_id>/edit/", edit_recommendation, name="edit_recommendation"),  # NEW
```
can take out the new comments

DonnieBLT commented on website/views/user.py:
```diff
@@ -31,8 +32,10 @@
 from blt import settings
-from website.forms import MonitorForm, UserDeleteForm, UserProfileForm
-from website.models import (
+from website.forms import RecommendationBlurbForm  # NEW - additive only
```
remove new comments

DonnieBLT reviewed (changes requested):
LGTM just a few nps

Pritz395 committed:
Remove 'NEW' comments and consolidate migrations

- Remove all 'NEW - additive only' comments from imports, docstrings, and code
- Consolidate migrations 0258, 0259, and 0264 into single 0258_add_recommendations.py
- Fix related_name in migration to match model definition

Pritz395 committed:
Refactor: Extract skills_by_category logic into helper function

- Create build_skills_by_category() helper to eliminate code duplication
- Replace 5 duplicated code blocks with helper function calls
- Change logger.error to logger.exception for better error tracking
- Reduces code by 34 lines and improves maintainability

Pritz395 commented:
> LGTM just a few nps

All done!

Pritz395 committed:
fix: Add database constraint to prevent self-recommendations and validate skills_endorsed

Pritz395 committed:
Renumber migrations to 0264 and 0265 to resolve conflict with main

Pritz395 committed:
Merge branch 'main' into feature/user-recommendations

DonnieBLT committed:
Merge branch 'main' into feature/user-recommendations

Pritz395 committed:
Merge branch 'main' into feature/user-recommendations

Pritz395 committed:
Merge branch 'main' into feature/user-recommendations

Pritz395 committed:
fix: Add hide action handler and remove dead code in recommendations

- Add hide/unhide functionality for approved recommendations
  * Toggle is_visible field on approved recommendations only
  * Shows appropriate success messages for hide/unhide states
  * Prevents hiding pending recommendations (approve or delete first)

- Remove unreachable is_approved check in delete_recommendation
  * Second check after permission validation was dead code
  * First check at line 2214 already returns if approved
  * Cleaner code flow: check approved → verify permission → delete

Pritz395 committed:
fix: Add prefetch_related to prevent N+1 queries in list_recommendations

Pritz395 committed:
style: Format code with ruff

Pritz395 committed:
fix: Handle missing profiles, existing requests, and safe error messages

Pritz395 committed:
style: Format code with ruff

Pritz395 committed:
fix: Update notification link to point to requester profile

Nachiket-Roy reviewed (changes requested):
Squash migrations into one coherent migration.
@coderabbitai could we use some indexes in this scenario? also check for missing rate limit, cache stampede or related caching issues, database constraint , json to manytomany and race condition.

Nachiket-Roy commented:
> Squash migrations into one coherent migration. @coderabbitai could we use some indexes in this scenario? also check for missing rate limit, cache stampede or related caching issues, database constraint , json to manytomany and race condition.

@coderabbitai

DonnieBLT committed:
Merge branch 'main' into feature/user-recommendations

DonnieBLT commented:
we are migrating away from Django and will revisit this feature then.
